Added New tutorials for Linea (#927)
Consensys/doc.linea: docs/learn/marketplace-dapp/marketplace-dapp.mdx

> # Building a Simple Marketplace DApp on Linea: A Step-by-Step Guide

Reviewer: We need front matter here, e.g. 

```
---
title: Marketplace dapp
description: A step-by-step guide for building a simple marketplace dapp
---
```

^ This keeps the `title` short to help with display in the sidebar, and moves some of the detail to the `description`. It also means we no longer need the `#` title you have line 1

### Introduction

In web3 ecosystem, dectralized applications (DApps) has become quite popular because of the ability to provide secure, transparent and trustless service. One of the most popular types of DApps are decentralized marketplaces to give the users ability to buy and sell items directly on the blockchain while removing the need of any middlemen.

In this tutorial, we will build a simple marketplace DApp on Linea, a layer 2 zkEVM technology which provides a scalable and cost-efficient environment while having the security of Ethereum. 

We will cover both the **frontend** and **smart contract development.**

> ## **What is Linea L2 and zkEVM?**

Author: This totally makes sense. Thank you!

INFURA_API_KEY=your_infura_api_key_here

# Private key of the account to be used for deployments and transactions
ACCOUNT_PRIVATE_KEY=your_account_private_key_here

```

### Deploy the Smart Contract

To deploy the smart contract to the Linea testnet, run the following command:

```bash
npx hardhat ignition deploy ignition/modules/Marketplace.ts --network linea-testnet

```

This will deploy the marketplace smart contract on Linea Sepolia. You can always write a script to deploy quicker by writing small commands like this on `package.json` file:

```jsx
"deploy:testnet": "npx hardhat ignition deploy ignition/modules/Marketplace.ts --network linea-testnet"
```

After the deployment we’ll get the `contract address` , let’s keep that. We will need it soon while integrating with the frontend. 

## Frontend: Next.js with Shadcn UI

### Create and Navigate to the Project Directory

```bash
mkdir site
cd site

```

### Initialize Next.js Project

```bash
npx create-next-app@latest .

```

When prompted, choose the following options:

- TypeScript: Yes
- ESLint: Yes
- Tailwind CSS: Yes
- `src/` directory: No (or Yes, if you prefer)
- App Router: Yes

> - Import alias: Yes (default @/*)

Author: It doesn't seem commented on my end. @jlwllmr

export const ABI = 
  // Paste the ABI here
```

Remember the deployed contract that we’ve saved earlier, we’ll add it here. Head over to `blockchain/ignition` module and find the `artifacts` folder to find the ABI.

![image](https://hackmd.io/_uploads/HJqu3tn20.png)

Let’s break down the code `app.tsx` and explain its key aspects in more detail:

## React and Hooks Usage

```jsx
import { useState, useEffect } from "react";
import { useAccount, useWalletClient } from "wagmi";

```

- The component uses React's `useState` for local state management and `useEffect` for side effects.
- It also uses custom hooks from `wagmi` (`useAccount` and `useWalletClient`) for blockchain wallet integration.

## State Management

```jsx
const [items, setItems] = useState<any[]>([]);
const [ownedItems, setOwnedItems] = useState<any[]>([]);
const [newItemName, setNewItemName] = useState("");
const [newItemPrice, setNewItemPrice] = useState("");

```

- Multiple state variables are defined to manage the component's data.
- `items` and `ownedItems` are arrays to store marketplace items.
- `newItemName` and `newItemPrice` are for form inputs when listing a new item.

> ## useEffect for Data Loading

Reviewer: ```suggestion
## `useEffect` for data loading
```

```jsx
useEffect(() => {
  loadItems();
  loadOwnedItems();
}, []);

```

- This effect runs once when the component mounts.
- It calls `loadItems()` and `loadOwnedItems()` to populate the state with data from the blockchain.

> ## Smart Contract Interaction

Reviewer: ```suggestion
## Smart contract interaction
```

```jsx
const loadItems = async () => {
  try {
    const itemCount = await client.readContract({
      address: CONTRACT_ADDRESS,
      abi: ABI_STRING_ARRAY,
      functionName: "itemCount",
    });
    // ... (fetching individual items)
  } catch (error) {
    console.error("Error loading items:", error);
  }
};

```

- This function reads data from the smart contract using `client.readContract`.
- It first gets the total item count, then fetches details for each item.

> ## Writing to the Blockchain

Reviewer: ```suggestion
## Writing to the blockchain
```

```jsx
const listItem = async () => {
  try {
    if (!walletClient) return;
    const { request } = await client.simulateContract({
      // ... contract call details
    });
    await walletClient.writeContract(request);
    loadItems();
  } catch (error) {
    console.error("Error listing item:", error);
  }
};

```

- This function writes data to the blockchain (listing a new item).
- It first simulates the contract call, then uses `walletClient.writeContract` to execute the transaction.

> ## UI Components and Styling

Reviewer: ```suggestion
## UI components and styling
```

```jsx
<Card className="p-4 sm:p-6" key={index}>
  <li key={item.id} className=" p-4">
    <p><strong>Name:</strong> {item.name}</p>
    {/* ... other item details */}
    <Button
      variant="outline"
      onClick={() => purchaseItem(item.id, item.price)}
      className="border-2 border-green-500 text-green-500 hover:bg-green-500 hover:text-white py-2 px-4 rounded  duration-200 hover:shadow-xl"
    >
      Purchase
    </Button>
  </li>
</Card>

```

- The component uses custom UI components like `Card` and `Button`.
- Tailwind CSS classes are used for styling (`className` props).
- Conditional rendering is used to show/hide the purchase button based on item status and ownership.

> ## Form Handling

Reviewer: ```suggestion
## Form handling
```

```jsx
<Input
  type="text"
  placeholder="Item Name"
  value={newItemName}
  onChange={(e) => setNewItemName(e.target.value)}
  className="border p-2 flex-1"
/>

```

- Controlled inputs are used for the form fields.
- The `value` and `onChange` props connect the input to the component's state.

> ## Error Handling

Reviewer: ```suggestion
## Error handling
```

```jsx
try {
  // ... contract interaction
} catch (error) {
  console.error("Error loading items:", error);
}

```

- Try-catch blocks are used throughout the code to handle potential errors in asynchronous operations, especially during blockchain interactions.

The `app.tsx` can become quite large so we can also refactor the code in`site/src/app/components/` directory.

```
/components
  - ListItem.tsx
  - AvailableItems.tsx
  - OwnedItems.tsx
/hooks
  - useItems.ts

```

> View the [full-code](https://github.com/meowyx/web3-marketplace-linea/blob/code-refactor/packages/site/src/app/page.tsx) here.

Reviewer: ```suggestion
View the [full code](https://github.com/meowyx/web3-marketplace-linea/blob/code-refactor/packages/site/src/app/page.tsx) here.
```

"complete source code"?

Also view the refactored code [here](https://github.com/meowyx/web3-marketplace-linea/tree/code-refactor/packages/site/src/app).

This component demonstrates advanced React patterns, integration with blockchain technology, and modern UI practices. It showcases how to build a decentralized application frontend that interacts with a smart contract while providing a user-friendly interface.

> ### Run the Development Server

Reviewer: ```suggestion
### Run the development server
```

```bash
npm run dev

```

Your Next.js application with Shadcn UI should now be running at `http://localhost:3000`.

> ![image](https://hackmd.io/_uploads/rkRw6YhhC.png)

Reviewer: As before, please amend to the HTML format

> Now we should be able to list and item,buy and transfer ownership.

Reviewer: ```suggestion
Now we should be able to list an item, buy and sell, and transfer ownership.
```

> In this guide, we built a simple decentralized marketplace DApp on Linea, leveraging zkEVM technology for scalability and cost efficiency. We covered everything from setting up the environment to writing and deploying smart contracts, and integrating the frontend with Next.js and ShadcnUI.

Reviewer: I think we can remove this paragraph

> As we continue to explore and expand on this foundational knowledge, consider the potential for further enhancements, such as integrating additional features or optimizing the user experience, bidding on items, auctions etc. The possibility of decentralized applications (DApps) are endless.

Reviewer: ```suggestion
As we continue to explore and expand on this foundational knowledge, consider the potential for further enhancements, such as integrating additional features or optimizing the user experience, bidding on items, auctions etc. The possibilities of dapps are endless.
```

# Building a Simple Marketplace DApp on Linea: A Step-by-Step Guide

> ### Introduction

Reviewer: ```suggestion
## Introduction
```

We jump immediately to a h3 (`###`) from h1 (`#`). Since we should only ever move in increments of 1, this should be a h2

In web3, decentralized applications (dapps) are popular because of their ability to provide secure, transparent, and trustless services. One of the most popular types is decentralized marketplaces, which enable users to buy and sell items directly on the blockchain while removing the need for any middlemen.

In this tutorial, we will build a simple marketplace dapp. 

We will cover both the **frontend** and **smart contract development.**

> 

Reviewer: We can remove these lines; we only need one line of space
